src/Generator/Generators/CSharp/CSharpTypePrinter.cs

```diff
@@ -36,7 +36,7 @@ public override TypePrinterResult VisitTagType(TagType tag, TypeQualifiers quals
                 return string.Empty;
 
             TypeMap typeMap;
-            if (TypeMapDatabase.FindTypeMap(tag, out typeMap))
+            if (TypeMapDatabase.FindTypeMap(tag, GeneratorKind.CSharp, out typeMap))
```

Comment:
Here we explicitly specify `GeneratorKind.CSharp` because 'CSharpTypePrinter' is used in CLI. Many similar cases are found in this file.

Reply:
After a bit of analysis and tests I found out that the CLI backend depends on CSharp typemaps in some ways.
Consider the `DelegatesPass` which is common for CLI and CSharp, it uses `CSharpTypePrinter` which itself uses the `CSharp` typemaps. Not passing the `GeneratorKind.CSharp` explicitly will make the `DelegatesPass` uses CLI typemaps, which is not desirable in this case.
Consider the [Employee.h](https://github.com/mono/CppSharp/blob/main/tests/dotnet/CLI/UseTemplateTypeFromIgnoredClassTemplate/Employee.h) example:
Using CLI typemap (undesired):
```C++
// Employee.h CLI generated header
[::System::Runtime::InteropServices::UnmanagedFunctionPointer(::System::Runtime::InteropServices::CallingConvention::Cdecl)]
delegate ::System::String^ Func__System_String^ ___IntPtr(::System::IntPtr __instance);
```
Using C# typemaps (desired):
```C++
// Employee.h CLI generated header
[::System::Runtime::InteropServices::UnmanagedFunctionPointer(::System::Runtime::InteropServices::CallingConvention::Cdecl)]
delegate ::System::String^ Func_std_basic_string___Internalc__N_std_S_basic_string__C___N_std_S_char_traits__C___N_std_S_allocator__C___IntPtr(::System::IntPtr __instance);
```
**Apparently this signature must match the `CSharp` one (in the generated .cs file) in order for the project to compile.**
To he honest, I was a bit confused why such mess is going on, but then I focused on refactoring rather than changing semantics.
This is why I am working on refactoring the project into modular design because with its current status it is nearly impossible to afford more generators with new semantics/heuristics. Isolating generators is a big win for future development, allowing reusable components as the current status which is also good (but we should add remarks/best practices).
Maybe someone with deeper knowledge on how the `CSharp/CLI` backend is operating can help us too. I can invest some time into these backends and see how they can be improved (and you should know that there is a lot of room for improvements), but I am thinking of sparing my efforts for newer generators.
By the way, I was thinking if we can standardize/document the typemap concept in details because it is a major concept in binding generation and currently I am finding it a bit confusing when comparing its usage across different generators. Maybe a kind of README/pseudocode can help a lot for users/collaborators.